mantle/cmd/ore/gcloud/promote-image.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 Red Hat
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcloud
+
+import (
+	"github.com/coreos/mantle/platform/api/gcloud"
+	"github.com/spf13/cobra"
+	"golang.org/x/net/context"
+)
+
+var (
+	cmdPromoteImage = &cobra.Command{
+		Use:   "promote-image",
+		Short: "Promote GCP image in image family",
+		Long:  "Promote GCP image in image family and deprecate all others",
+		Run:   runPromoteImage,
+	}
+
+	promoteImageName   string
+	promoteImageFamily string
+)
+
+func init() {
+	cmdPromoteImage.Flags().StringVar(&promoteImageName, "image", "", "GCP image name")
+	cmdPromoteImage.Flags().StringVar(&promoteImageFamily, "family", "", "GCP image family to promote within")
+	GCloud.AddCommand(cmdPromoteImage)
+}
+
+func deprecateImage(name string, state gcloud.DeprecationState, replacement string) {
+	plog.Infof("Changing deprecation state of image: %v -> %v",
+		name, string(state))
+	pending, err := api.DeprecateImage(name, state, replacement)
+	if err == nil {
+		err = pending.Wait()
+	}
+	if err != nil {
+		plog.Fatalf("Changing deprecation state of image failed: %v\n", err)
+	}
+}
+
+func runPromoteImage(cmd *cobra.Command, args []string) {
+	// Check that the user provided an image
+	if promoteImageName == "" {
+		plog.Fatal("Must provide an image name via --image")
+	}
+
+	plog.Infof("Attempting to promote %v in family %v",
+		promoteImageName, promoteImageFamily)
+
+	// Get all images in the image family
+	images, err := api.ListImages(context.Background(), "", promoteImageFamily)
+	if err != nil {
+		plog.Fatal(err)
+	}
+
+	// Make sure the specified image exists in the specified image family
+	found := false
+	for _, image := range images {
+		if image.Name == promoteImageName {
+			found = true
+		}
+	}
+	if !found {
+		plog.Fatalf("The image (%v) must be in the image family (%v)",
+			promoteImageName, promoteImageFamily)
+	}
+
+	// First undeprecate the image we want to promote
+	deprecateImage(promoteImageName, gcloud.DeprecationStateActive, "")
```

Comment:
Question: does the `api.DeprecateImage` call no-op without an error if the image is already in the desired deprecation state?

If not we might need to build some annoying checks that skips an action if it's already in the correct state so that this command can be re-ran if it errors the first time through.

Reply:
my experience is that it just returns successfully it it's already in the desired state, though as @bgilbert mentions in https://github.com/coreos/coreos-assembler/pull/1456#discussion_r426782096 we should try to minimize the number of API calls we do by only requesting to change the deprecation state for images that need it.

Reply:
ok updated the code so that it only calls the deprecate for images that need it